hips/hip-00xx-sequencing.md

```diff
@@ -0,0 +1,192 @@
+---
+hip: 9999
+title: "Better Support for Resource Creation Sequencing"
+authors: [ "Joe Beck <joebeck5705@gmail.com>", "Evans Mungai <mbuevans@gmail.com>" ]
+created: "2024-11-15"
+type: "feature"
+status: "draft"
+---
+
+## Abstract
+
+This HIP is to propose a new featureset in Helm 4 to provide Application developers, who create heam charts for their applications, a well supported way of defining what order chart resources and chart dependencies should be deployed to Kubernetes.
+
+## Motivation
+
+Today, to accomplish resource sequencing you have two options. The first is leveraging helm hooks, the second is building required sequencing into the application (via startup code or init containers). The existing hooks and weights can be tedious to build and maintain for Application developers, and built-in app sequencing can unecessarily increase complexity of a Helm application that needs to be maintained by Application Developers. Helm as a package manager should be capable of enabling developers to properly sequence how their applications are deployed, and by providing such a mechanism to developers, this will significantly improve the Application developer's experience.
+
+Additionally, Helm currently doesn't provide a way to sequence when chart dependencies are deployed, and this featureset would ideally address this.
+
+## Rationale
+
+### Proposal 1: Named Dependencies
```

Comment:
Personally, I quite like the general direction of proposal 1—as a replacement of numeric weights entirely—over proposal 2, which extends that. My point of view is I would like to move the numeric weighting into the "alternatives considered" section. But would like to hear other maintainers and community feedback on this.

I have wanted to introduce a depends_on functionality for helm resources—as a dynamic alternative to weights—to allow users to specify a specific order they are applied (a similar concept is used by used by projects like terraform, flux, and docker-compose).

My main reasoning for this is numbered weights are static—you must know the entire list of how all resources should be ordered, and this often takes a lot of finagling and fussing to get it right.

While named dependencies are dynamic—creating a DAG on the fly, based only on when and if certain resources have an actual dependency on one or more other resources.

It might be good to note some of this in the proposal. I could sketch up some suggested text for this if you prefer, but would also be happy for you to take a stab at it in your HIP. What do you think?

Reply:
Go ahead! I'll add you as a contributor, I appreciate the assistance.

Reply:
I moved the weights proposal to rejected ideas and tried to capture some of the points you made here in why.